### PR TITLE
SPARKC-198: Meaningful exception if some partition key column is null

### PR DESCRIPTION
**Pull request by jacek-lewandowski**

### CODE REVIEW

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala [hunk 1]
```diff
@@ -44,8 +44,12 @@ class RoutingKeyGenerator(table: TableDef, columnNames: Seq[String])
 
   private def fillRoutingKey(stmt: BoundStatement): Array[ByteBuffer] = {
     val rk = routingKey.get
-    for (i <- 0 until partitionKeyIdxs.size)
+    for (i ← partitionKeyIdxs.indices) {
+      if (stmt.isNull(partitionKeyIdxs(i)))
+        throw new NullPointerException(
```
Comment (helena) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala]: IMO never throw an NPE, in fact, why do you do _any_ work prior to checking the argument to the function?

        private def fillRoutingKey(stmt: BoundStatement): Array[ByteBuffer] = {
           require(stmt != null, "BoundStatement 'stmt' must have a valid value.") 
        }

then in your specs check for IllegalArgumentException, because really you are trying to alert the engineer in runtime that you can't pass nulls to this api function. If this is called by our api then we'd better be checking the arg state!

File: spark-cassandra-connector/src/it/scala/com/datastax/spark/connector/rdd/RDDSpec.scala [hunk 2]
```diff
@@ -201,6 +201,38 @@ class RDDSpec extends SparkCassandraITFlatSpecBase {
     checkArrayCassandraRow(result)
   }
 
+  it should "throw a meaningful exception if partition column is null when joining with Cassandra table" in {
+    val source = sc.parallelize(keys).map(x ⇒ new KVWithOptionRow(None))
+    try {
+      source.joinWithCassandraTable[(Int, Long, String)](ks, tableName).collect()
+      fail()
```
Comment (helena) [inline spark-cassandra-connector/src/it/scala/com/datastax/spark/connector/rdd/RDDSpec.scala]: I'm not clear why you are manually adding fail() shouldn't you be testing args that are null/None and the changed code you wrote be throwing the failure to note the args were not valid? In which case you should be doing

intercept[IllegalArgumentException] {
  source.doThingsThatShouldFail()
}

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala [hunk 3]
```diff
@@ -7,7 +7,7 @@ import com.datastax.spark.connector.cql.TableDef
 
 /** This class computes the routing key of a bound statement. */
 class RoutingKeyGenerator(table: TableDef, columnNames: Seq[String])
-  extends ((BoundStatement) => ByteBuffer) {
+  extends ((BoundStatement) ⇒ ByteBuffer) {
```
Comment (helena) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala]: Did you add scala formatting? Unclear why you added ⇒ vs =>, ⇒ is what formatting adds

File: spark-cassandra-connector/src/it/scala/com/datastax/spark/connector/rdd/RDDSpec.scala (hunk 2 above)
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/it/scala/com/datastax/spark/connector/rdd/RDDSpec.scala]: As you can see, I need to verify the exception message - that is why I did it in this way

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala (hunk 1 above)
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala]: Actually this is NPE because the argument is null. I'm not sure about IAE because I think this should be used for method arguments.
Also we cannot identify them sooner because they are in RDD, so in the distributed data. They are not processed in the driver.

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala (hunk 3 above)
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala]: IntelliJ does this automatically if you type => + space

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala (hunk 1 above)
Comment (helena) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala]: I don't understand why you have to throw a NPE, why not require like my sample, I get that it's in runtime while parsing partitionKeyIdxs

Comment (helena) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala]: And if you need to validate the message in tests, why not create a custom exception and test the exception type vs strings

Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala]: I can change to IAE. However, how introducing a custom exception would allow me to verify the exception message which should say which column is null?

Comment (helena) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala]: What about:

         throw new NullOrEmptyColumnException(s"Invalid null value for partition key part $theValue")

         @SerialVersionUID(1L)
          class NullOrEmptyColumnException(columnName: String, message: String, cause: Throwable) extends             RuntimeException(message, cause) {

               def this(columnName: String) = this(columnName,"", null) 
           }

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala (hunk 3 above)
Comment (helena) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala]: Ug, I've never seen that, not good that it does that.

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala (hunk 1 above)
Comment (helena) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala]: Then we can do this and check type vs string

       intercept[InvalidColumnException] {
           source.doThingsThatShouldFailOrThisTestShouldFail()
       }

WDYT?

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala (hunk 3 above)
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala]: Why? don't you like that nice arrows? :)

File: spark-cassandra-connector/src/it/scala/com/datastax/spark/connector/rdd/RDDSpec.scala [hunk 4]
```diff
@@ -201,6 +201,38 @@ class RDDSpec extends SparkCassandraITFlatSpecBase {
...
+    } catch {
+      case ex: Exception ⇒
+        ex.getMessage should include("Invalid null value for partition key part key")
```
Comment (pkolaczk) [inline spark-cassandra-connector/src/it/scala/com/datastax/spark/connector/rdd/RDDSpec.scala]: This way you're not testing if the exception has been thrown at all. Better use the pattern recommended by scalatest: 
val e = the [Exception] thrown by { some code }
e.getMessage should ...

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala (hunk 3 above)
Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala]: I think this is something we must agree upon. So far, my IntelliJ does not do that, and we have nice arrows in half of the places and ugly arrows in another half.

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala (hunk 1 above)
Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala]: This is a good idea with a custom exception class.

Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/RoutingKeyGenerator.scala]: I don't like, but I don't want to fight for not creating it. Basically because it gives nothing but another class. The user is unable to catch it because it is thrown in executor. The driver will only get `SparkException` with the message but without the cause. So such class is useless.

**Comment by helena** [comment]
LGTM thanks

**Comment by jacek-lewandowski** [comment]
test this please
